src/orange/detail/BIHBuilder.cc

```diff
@@ -0,0 +1,225 @@
+
+namespace celeritas
+{
+namespace
+{
+//---------------------------------------------------------------------------//
+/*!
+ * Variadic-templated struct for overloaded operator() calls.
+ */
+template<typename... Ts>
+struct Overload : Ts...
+{
+    using Ts::operator()...;
+};
+
+//---------------------------------------------------------------------------//
+/*!
+ * "Deduction guide" for instantiating Overload objects w/o specifying types.
+ */
+template<class... Ts>
+Overload(Ts&&...) -> Overload<Ts...>;
+}  // namespace
+namespace detail
+{
+//---------------------------------------------------------------------------//
+/*!
+ * Construct from a Storage object
+ */
+BIHBuilder::BIHBuilder(BIHStorage storage) : storage_(storage) {}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Create BIH Nodes.
+ */
+BIHTree BIHBuilder::operator()(VecBBox bboxes)
+{
+    // Store bounding boxes and their corresponding centers
+    CELER_EXPECT(!bboxes.empty());
+    bboxes_ = std::move(bboxes);
+    centers_.resize(bboxes_.size());
+    std::transform(bboxes_.begin(),
+                   bboxes_.end(),
+                   centers_.begin(),
+                   &celeritas::center<fast_real_type>);
+
+    VecIndices indices;
+    VecIndices inf_volids;
+
+    for (auto i : range(bboxes_.size()))
+    {
+        LocalVolumeId id(i);
+
+        if (!is_infinite(bboxes_[i]))
+        {
+            indices.push_back(id);
+        }
+        else
+        {
+            inf_volids.push_back(id);
+        }
+    }
+
+    VecNodes nodes;
+    this->construct_tree(indices, &nodes, BIHNodeId{});
+
+    auto [inner_nodes, leaf_nodes] = this->arrange_nodes(std::move(nodes));
+
+    BIHTree params;
+
+    params.bboxes = ItemMap<LocalVolumeId, FastBBoxId>(
+        make_builder(storage_.bboxes)
+            .insert_back(bboxes_.begin(), bboxes_.end()));
+
+    params.inner_nodes
+        = make_builder(storage_.inner_nodes)
+              .insert_back(inner_nodes.begin(), inner_nodes.end());
+    params.leaf_nodes = make_builder(storage_.leaf_nodes)
+                            .insert_back(leaf_nodes.begin(), leaf_nodes.end());
+
+    params.inf_volids = make_builder(storage_.local_volume_ids)
+                            .insert_back(inf_volids.begin(), inf_volids.end());
+
+    return params;
+}
+
+//---------------------------------------------------------------------------//
+// HELPER FUNCTIONS
+//---------------------------------------------------------------------------//
+/*!
+ * Recursively construct BIH nodes for a vector of bbox indices.
+ */
+void BIHBuilder::construct_tree(VecIndices const& indices,
+                                VecNodes* nodes,
+                                BIHNodeId parent) const
+{
+    auto current_index = nodes->size();
+    nodes->resize(nodes->size() + 1);
+
+    BIHPartitioner partitioner(&bboxes_, &centers_);
+    auto p = partitioner(indices);
+
+    if (p)
+    {
```

Comment:
Function-like objects should be verbs, just like functions.

```suggestion
    BIHPartitioner partition(&bboxes_, &centers_);

    if (auto p = partition(indices))
    {
```